src/services/alarm-chat.service.js

```diff
@@ -0,0 +1,99 @@
+const { alarmService, ALARM_EVENT_KEY } = require('./alarm.service');
+const { redisService } = require('./redis.service');
+const { getAlarmStartNotificationMessage, alarmEndNotificationMessage, chatIsMutedMessage, chatIsUnmutedMessage } = require('../message');
+const { handleError } = require('../utils');
+
+class AlarmChatService {
+  /**
+   * @param {GrammyBot['api']} api
+   * */
+  async init(api) {
+    this.api = api;
+    this.chats = await this.getChatsWithAlarmModeOn();
+    this.subscribeToAlarms();
+  }
+
+  /**
+   * @param {ChatSessionData} chatSession
+   * @param {string} id
+   * */
+  async updateChat(chatSession, id) {
+    const chatId = id.toString();
+    const index = this.chats.findIndex((chat) => chat.id === chatId);
+    if (index !== -1) {
+      if (!chatSession.chatSettings.disableChatWhileAirRaidAlert && !chatSession.chatSettings.airRaidAlertSettings.notificationMessage) {
+        this.chats.splice(index, 1);
+      } else {
+        this.chats[index].data = chatSession;
+      }
+    } else if (chatSession.chatSettings.disableChatWhileAirRaidAlert || chatSession.chatSettings.airRaidAlertSettings.notificationMessage) {
+      this.chats.push({
+        id: chatId,
+        data: chatSession,
+      });
+    }
+  }
+
+  /**
+   * @returns {Promise<ChatSession[]>}
+   * */
+  async getChatsWithAlarmModeOn() {
+    const sessions = await redisService.getChatSessions();
+    return sessions.filter(
+      (s) => s.data.chatSettings?.airRaidAlertSettings?.notificationMessage || s.data.chatSettings?.disableChatWhileAirRaidAlert,
+    );
+  }
+
+  subscribeToAlarms() {
+    alarmService.updatesEmitter.on(ALARM_EVENT_KEY, (event) => {
+      this.chats.forEach(async (chat) => {
+        if (chat.data.chatSettings.airRaidAlertSettings.state === event.state.name) {
+          await this.processChatAlarm(chat, event.state.alert);
+        }
+      });
```

Comment:
This is a very bad way to do async tasks:

1) forEach with async is ineffective because if the error occurs, our error handling won't handle it. Also, it will be called all at once which could create significant API calls per sec. We could meet a rate limit; You can use Promise.all (not suitable here) or for of (not suitable as well);
https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop

2) It's better to use Queue here because we have rate limits. You can try to use a queue that I used for global updates.